### PR TITLE
JP-3593: Fix 2nd/3rd group saturation for NIRSpec IRS2

### LINKED ISSUE #8413: Saturation step should increase flagging for group 3 saturation
_Issue [JP-3593](https://jira.stsci.edu/browse/JP-3593) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

*Edited based on discussion thread below.*

Examination of a few extremely-large outliers in PID 1908 (NIRSpec IFU) indicates that these are being caused by cosmic rays that are not being properly flagged in the jump step.

In this case, ramps are 22 groups long.  Since the data is grouped though (nframes/group=5) cosmic rays that cause saturation in group 3 are also significantly biasing group 2 half-way to saturation.  Since all other groups are saturated the pipeline assumes that groups 1 and 2 are good, resulting in extremely bright ramps.

This isn't as much of a problem when saturation is first flagged in group 4 or greater, as there are multiple differences with which to find and reject the abnormally large slope.  Likewise, it's not an issue when saturation is first flagged in group 2, as this simply results in no slope at all.  A solution can thus concentrate specifically on the case of saturation in group 3.

At the same time, we want to be careful not to introduce excessive flagging when saturation occurs due to normal counts received from a source, as this would decrease the effective dynamic range of some observations.

The proposed solution is therefore to modify the saturation step.  For any data obtained with nframes/group > 1, look for pixels for which SATURATION is set in group 3.  For each of these pixels, look at the counts in group 1.  If the count value in group 1 is less than 1/10 of the value that would be required to trigger SATURATION (i.e., it isn't obviously a very bright source), then also set the SATURATION flag for this pixel in group 2.

Once this has been coded it will be necessary for testing to occur on both NIRSpec IFU data to ensure that it fixes the original problem, and on NIRISS/NIRCAM data to ensure that it doesn't produce any unintended consequences.

### PR DESCRIPTION
**Pull request by drlaw1558**
Resolves [JP-3593](https://jira.stsci.edu/browse/JP-3593) by adding in a set of checks for pixels that saturate in the 3rd group that already saturated in the 2nd group but were diluted by the group average.  Only affects NIRSpec IRS2 data.

Closes #8413 

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [x] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by hbushouse** [comment]
Started a regtest run at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1559

**Comment by hbushouse** [comment]
@drlaw1558 What about RTD updates? Probably should mention this detail in the saturation step docs.

**Comment by hbushouse** [comment]
> Started a regtest run at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1559

I've confirmed that all ~200 failures in the regtest run are unrelated and due to the use of an older version of stdatamodels in the PR branch. So this looks OK.

File: docs/jwst/saturation/description.rst [hunk 1]
```diff
@@ -34,6 +34,13 @@ The "NO_SAT_CHECK" flag is propagated to the
 PIXELDQ array in the output science data to indicate which pixels fall into
 this category.
 
+For grouped data, this step also looks for and flags data that saturated during
+the second group but did not trigger the normal saturation threshold due to the
+grouped data averaging.  This requires that the third group be saturated, and
+the first group sufficiently low that the third group would not have been expected
```
Comment (hbushouse) [inline docs/jwst/saturation/description.rst:40]: Not sure I understand this. This only catches cases where group 3 must've been affected by a CR hit, which then caused it to be above the saturation threshold? But if it's the case that the high level in group 3 was due to a CR hit that occurred somewhere in group 3, why is group 2 then suspect? Or are you saying that the CR hit actually occurred in one of the frames of group 2, but did not drive the average for group 2 above the saturation threshold?

Comment (drlaw1558) [reply docs/jwst/saturation/description.rst:40]: The latter.  CR hit actually caused saturation in group 2, but the averaging kept it below threshold.

File: jwst/saturation/saturation.py [hunk 2]
```diff
@@ -160,6 +162,31 @@ def irs2_flag_saturation(input_model, ref_model, n_pix_grow_sat):
...
+
+                # Identify groups that are saturated in the third group
+                gp3mask = np.where(flag_temp & SATURATED, True, False)
+                mask &= gp3mask
+
+                # Flag the 2nd group for the pixels passing that gauntlet in the 3rd group
+                dq_temp = np.zeros_like(mask,dtype='uint8')
+                dq_temp[mask] = SATURATED
+                # flag any pixels that border saturated pixels
+                if n_pix_grow_sat > 0:
```
Comment (hbushouse) [inline jwst/saturation/saturation.py:183]: Is it really worthwhile or necessary to extend the flags for this situation to neighboring pixels, given that the group 2 pixels are only barely saturated to begin with? Is there much of an effect on the neighbors?

Comment (drlaw1558) [reply jwst/saturation/saturation.py:183]: Yup.  See attached for an example; left is the B10.2 rate image, right is after this PR.  Most pixel values are about 1e-3 DN/s, the core of the hit is 493 DN/s, and the ring around it is anywhere from 3-100 DN/s (still orders of magnitude above the non-affected pixels).
Screenshot 2024-06-25 at 12 54 17 PM

Comment (hbushouse) [reply jwst/saturation/saturation.py:183]: Wow, OK. So the answer is a definite "yes"!

**Review by hbushouse — APPROVED** [review]
All test results are good, all questions/comments addressed, and code looks good.

**Pull request merged**